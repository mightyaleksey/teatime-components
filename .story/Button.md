## Preview


### theme action

```
<div>
  <Button
    disabled={true}
    size='s'
    theme='action'>Action Button</Button>
  {' '}size s
  <Button
    size='s'
    theme='action'>Action Button</Button>
  {' '}size m
  <Button
    size='m'
    theme='action'>Action Button</Button>
  {' '}size l
  <Button
    size='l'
    theme='action'>Action Button</Button>
</div>
```


### theme link

```
<div>
  <Button
    disabled={true}
    size='s'
    theme='link'>Link Button</Button>
  {' '}size s
  <Button
    size='s'
    theme='link'>Link Button</Button>
  {' '}size m
  <Button
    size='m'
    theme='link'>Link Button</Button>
  {' '}size l
  <Button
    size='l'
    theme='link'>Link Button</Button>
</div>
```


### theme normal

```
<div>
  <Button
    disabled={true}
    size='s'
    theme='normal'>Normal Button</Button>
  {' '}size s
  <Button
    size='s'
    theme='normal'>Normal Button</Button>
  {' '}size m
  <Button
    size='m'
    theme='normal'>Normal Button</Button>
  {' '}size l
  <Button
    size='l'
    theme='normal'>Normal Button</Button>
</div>
```
