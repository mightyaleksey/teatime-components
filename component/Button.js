'use strict';

const {Component, PropTypes} = require('react');
const {filterProps, genericName} = require('../lib/util');
const React = require('react');
const cc = require('classnames');

const cssModules = {
  'action-l': require('../style/button/button-action-l.css'),
  'action-m': require('../style/button/button-action-m.css'),
  'action-s': require('../style/button/button-action-s.css'),
  'link-l': require('../style/button/button-link-l.css'),
  'link-m': require('../style/button/button-link-m.css'),
  'link-s': require('../style/button/button-link-s.css'),
  'normal-l': require('../style/button/button-normal-l.css'),
  'normal-m': require('../style/button/button-normal-m.css'),
  'normal-s': require('../style/button/button-normal-s.css'),
};

class Button extends Component {
  css = tokenName => genericName(this.props, tokenName)

  focus() {
    if (!this._button) return;
    this._button.focus();
  }

  render() {
    const {autoFocus, children, className, disabled, id, type} = this.props;
    const {css} = this;

    return (
      <button
        {...filterProps(this.props)}
        autoFocus={autoFocus}
        className={cc(css('control'), className)}
        disabled={disabled}
        id={id}
        ref={r => this._button = r}
        type={type}>
        {children}
      </button>
    );
  }
}

Button.defaultProps = {
  size: 'm',
  styles: cssModules,
  theme: 'normal',
};

Button.propTypes = {
  size: PropTypes.oneOf([
    'l',
    'm',
    's',
  ]),
  styles: PropTypes.object,
  theme: PropTypes.oneOf([
    'action',
    'link',
    'normal',
  ]),
};

module.exports = Button;
