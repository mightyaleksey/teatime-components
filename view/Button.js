'use strict';

const { Component, PropTypes } = require('react');
const { styleName } = require('../tool/className');
const { sanitizeProps } = require('../tool/component');
const React = require('react');

class Button extends Component {
  focus() {
    if (this.refs.control) {
      this.refs.control.focus();
    }
  }

  filter(props) {
    let filteredProps = {};

    Object.keys(props).forEach(prop => {
      if (!Button.illegalDOMProps.includes(prop)) {
        filteredProps[prop] = props[prop];
      }
    });

    return filteredProps;
  }

  render() {
    return (
      <button {...sanitizeProps(['theme', 'styles', 'styleName'], this.props)} className={styleName(this.props)} ref='control'/>
    );
  }
}

Button.defaultProps = {
  styleName: 'control',
  styles: {},
};

Button.propTypes = {
  styleName: PropTypes.string,
  styles: PropTypes.shape({
    control: PropTypes.string,
  }),
};

module.exports = Button;
