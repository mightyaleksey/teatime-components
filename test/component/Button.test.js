import Button from '../../component/Button';
import React from 'react';
import renderer from 'react-test-renderer';

test('Button renders correctly', () => {
  const tree = renderer.create(
    <Button
      autoFocus={true}
      className='outer'
      disabled={true}
      id='buttons id'
      type='submit'>
      Normal Button
    </Button>
  ).toJSON();

  expect(tree).toMatchSnapshot();
});
