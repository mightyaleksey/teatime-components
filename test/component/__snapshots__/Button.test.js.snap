exports[`test Button renders correctly 1`] = `
<button
  autoFocus={true}
  className="control outer"
  disabled={true}
  id="buttons id"
  type="submit">
  Normal Button
</button>
`;
